### PR TITLE
Brokering to biostudies

### PR DESCRIPTION
**Pull request by ke4**
This branch is branched off from the `ena-conversion` branch. Some of those codes were not compiling, so I had to comment them. I suspect the latest changes in the `json-converter` or other dependencies were not committed properly (but probably worked locally).

Changes:
- Configure Travis CI for unit tests execution
- Reorganised tests
- Add brokering submissions into BioStudies archive
- Update BioStudies submission with links to other archive entities

TODO: There are some broken unit tests that need to be fixed. Sadly, they were already broken in that branch I branched off: `ena-conversion`.

### CODE REVIEW

**Comment by MightyAx** [comment]
I've fixed the conversion_utils.fixed_attribute reference.

File: cli.py [hunk 1]
```diff
@@ -22,6 +29,13 @@ class CovidExcelUtils:
     def __init__(self, file_path, output):
         self.__file_path = file_path
         self.__output = output
+        self.ena_submission_files = {}
+
+        self.bio_studies_service = BioStudies(
+            args['biostudies_url'],
+            os.environ['BIOSTUDIES_USERNAME'],
+            os.environ['BIOSTUDIES_PASSWORD']
+        )
```
Comment (MightyAx) [inline cli.py]: this breaks the cli if the environment variables aren't present.
The CovidExcelUtilsClass shouldn't use any environment variables directly, the main method below should convert from the os variables to keep the class implementation clean.
```
➜  cli.py examples/12_v3_raw_reads.xlsx --output json
Traceback (most recent call last):
  File "/Users/astaffer/.pyenv/versions/3.8.3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/astaffer/.pyenv/versions/3.8.3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/astaffer/.vscode/extensions/ms-python.python-2021.1.502429796/pythonFiles/lib/python/debugpy/__main__.py", line 45, in <module>
    cli.main()
  File "/Users/astaffer/.vscode/extensions/ms-python.python-2021.1.502429796/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 444, in main
    run()
  File "/Users/astaffer/.vscode/extensions/ms-python.python-2021.1.502429796/pythonFiles/lib/python/debugpy/../debugpy/server/cli.py", line 285, in run_file
    runpy.run_path(target_as_str, run_name=compat.force_str("__main__"))
  File "/Users/astaffer/.pyenv/versions/3.8.3/lib/python3.8/runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/astaffer/.pyenv/versions/3.8.3/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/astaffer/.pyenv/versions/3.8.3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/astaffer/Repos/covid-excel-utils/cli.py", line 208, in <module>
    with closing(CovidExcelUtils(args['file_path'], args['output'])) as excel_utils:
  File "/Users/astaffer/Repos/covid-excel-utils/cli.py", line 36, in __init__
    os.environ['BIOSTUDIES_USERNAME'],
  File "/Users/astaffer/.pyenv/versions/3.8.3/lib/python3.8/os.py", line 675, in __getitem__
    raise KeyError(key) from None
KeyError: 'BIOSTUDIES_USERNAME'

```

File: submission/entity.py [hunk 2]
```diff
@@ -49,10 +49,13 @@ def add_errors(self, attribute: str, error_msgs: Iterable[str]):
         self.errors.setdefault(attribute, []).extend(error_msgs)
         
     def add_link_id(self, identifier: EntityIdentifier):
-        self.links.setdefault(identifier.entity_type, []).append(identifier)
```
Comment (ke4) [inline submission/entity.py]: This change is going to break my current accession update code.
Although now I reviewed your whole change set and I can see that you changed also the accession handling.
So I suggest to do pair programming and finish the changes together.

File: submission/entity.py [hunk 3]
```diff
@@ -49,10 +49,13 @@ def add_errors(self, attribute: str, error_msgs: Iterable[str]):
...
+        self.add_link(identifier.entity_type, identifier.index)    
 
+    def add_link(self, entity_type: str, index: str):
```
Comment (ke4) [inline submission/entity.py:29]: is that method called anywhere else? If not then it should be a private method.
Also not clear to me why the `add_link_id` and `add_link` are two separate method.

File: conversion/ena/submission.py [hunk 4]
```diff
@@ -81,3 +81,23 @@ def runs_file(data: Submission) -> Element:
...
         return runs_node
+    
+    @staticmethod
+    def submission_file(data: Submission) -> Element:
+        submission = etree.Element('SUBMISSION')
+        actions = etree.SubElement(submission, 'ACTIONS')
+        action_add = etree.SubElement(actions, 'ACTION')
+        etree.SubElement(action_add, 'ADD')
+        if 'study' in data.get_entity_types():
+            release_dates = set()
```
Comment (ke4) [inline conversion/ena/submission.py:92]: I am afraid to set the release date on a study is not that easy.
https://ena-docs.readthedocs.io/en/latest/submit/general-guide/programmatic.html#submission-xml-set-study-hold-date
This means that 1st you have to submit that study AND THEN you have to do another submission to update its release date.

File: cli.py [hunk 5]
```diff
@@ -52,54 +47,41 @@ def validate(self):
...
-            bio_study_submission = BioStudyConverter.convert_study(bio_study)
-            accession = self.bio_studies_service.send_submission(bio_study_submission)
+        study: Entity
+        for study in self.excel.data.get_entities("study"):
+            bio_study_submission = BioStudyConverter.convert_study(study)
+            accession = service.send_submission(bio_study_submission)
             accessions.append(accession)
-            bio_study.attributes['bio_study_accession'] = accession
-
-        biostudies_accession_count = len(accessions)
```
Comment (ke4) [inline cli.py]: I would add back these lines to have some log from a biostudies submission

File: excel/load.py [hunk 6]
```diff
@@ -59,9 +64,14 @@ def get_data(worksheet, column_map: dict) -> ExcelSubmission:
                     attribute_name = column_map[cell.column_letter]['attribute']
                     row_data.setdefault(object_name, {})[attribute_name] = value
             for entity_type, attributes in row_data.items():
-                index = ExcelLoader.get_index(entity_type, row_index, attributes)
                 accession = ExcelLoader.get_accession(entity_type, attributes)
-                data.map_row(row_index, entity_type, index, accession, attributes)
+                if accession:
```
Comment (ke4) [inline excel/load.py:68]: It is not clear to me how would we know from this point whether an index contains an `accession` or an `alias` as both of them are string values.

File: services/biostudies.py [hunk 7]
```diff
@@ -3,11 +3,12 @@
 from biostudiesclient.api import Api
 from biostudiesclient.auth import Auth
 
+from submission.entity import Entity
+
 LINK_TYPE_MAP = {
-    'sample': 'BioSample',
+    'sample': 'BioSamples',
```
Comment (ke4) [inline services/biostudies.py:17]: That should be `BioSample` for this reason: https://github.com/EBIBioStudies/ribs/blob/4d0ea417fbd30fd325ddacec127a3e6940199faa/src/main/webapp/js/detail/detailpage.js#L93

File: submission/entity.py [hunk 8]
```diff
@@ -37,6 +32,17 @@ def add_link(self, entity_type: str, index: str):
...
-    def get_accession(self):
-        accession = self.identifier.accession
-        return accession
+    def add_accession(self, service: str, accession_value: str):
+        self.__accessions[service] = accession_value
+
+    def get_accession(self, service: str) -> str:
+        return self.__accessions.get(service, None)
+
+    def get_first_accession(self, service_priority):
```
Comment (ke4) [inline submission/entity.py:41]: What is the use case for this method?
I don't really understand why would this be useful.
Also: missing test case to be able to understand why we would need this

**Review by ke4 — COMMENTED** [review]
I added some comments. We could resolve them later today.
I am going to setup a meeting for this afternoon. Tx

File: conversion/ena/submission.py (hunk 4 above)
Comment (MightyAx) [reply conversion/ena/submission.py:92]: This was taken from here:
https://ena-docs.readthedocs.io/en/latest/submit/study/programmatic.html#create-the-submission-xml

Us discussed in person, let's ensure we verify this during ENA testing. (after we merge)
```submission.xml
<SUBMISSION>
   <ACTIONS>
      <ACTION>
         <ADD/>
      </ACTION>
      <ACTION>
         <HOLD HoldUntilDate="TODO: release date"/>
      </ACTION>
   </ACTIONS>
</SUBMISSION>
```

File: excel/load.py (hunk 6 above)
Comment (ke4) [reply excel/load.py:68]: We discussed this with Alexie and this should work for our case

**Pull request merged**